### PR TITLE
Support for asynchronous lambda invocation with integration type AWS

### LINKED ISSUE #4862: Add support for asynchronous lambda invocation with integration type AWS
# This is a Feature Proposal

## Description

API Gateway supports asynchronous lambda invocation when integration type is AWS. Described here:
https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-lambda.html
Currently it is not possible to configure this in serverless directly and requires a resource override to enable. This feature proposal is to add support for integration type AWS, asynchronous lambda invocation without using an resource override. 

* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

Asynchronous lambda invocation allows for a quick response back (e.g. 202 Accepted) while allowing the lambda to continue running. This would be desired if you wish to kick off a lambda that runs longer then the API GW 30 second timeout, or if the application needs a quick response without the need for direct success for failure of the longer running lambda.  

* If there is additional config how would it look

Add new integration type `AWS_ASYNC` that configures the `X-Amz-Invocation-Type` correctly:
```yaml
  lambda:
    handler: index.handler
    events:
      - http:
          path: users
          method: post
          integration: AWS_ASYNC
```
* Cloud Formation Information

`AWS_ASYNC` would set the integration type to AWS and add ` "integration.request.header.X-Amz-Invocation-Type": "'Event'"` to `Integration { RequestParameters: {} }` similar to as follows:
```
          "ApiGatewayMethodUsersPost": {
            "Type": "AWS::ApiGateway::Method",
            "Properties": {
              "HttpMethod": "POST",
              "RequestParameters": {
              },
              "Integration": {
                "IntegrationHttpMethod": "POST",
                "Type": "AWS",
                "RequestParameters": {
                  "integration.request.header.X-Amz-Invocation-Type": "'Event'"
                }
```
## Additional Data

* ***Serverless Framework Version you're using***:
serverless/serverless as of 4/28/2018
* ***Operating System***:
Ubuntu 16.04
* ***Stack Trace***:
N/A
* ***Provider Error messages***:
N/A

### PR DESCRIPTION
**Pull request by snurmine**
## What did you implement:

Closes #4862

## How can we verify it:

```yml
service: env-test
provider:
  name: aws
  runtime: nodejs6.10

functions:
   lambda:
    handler: index.handler
    events:
      - http:
          path: users
          method: post
          async: true
```
Running `sls package` should generate template that has 
```
 "RequestParameters": {
     "integration.request.header.X-Amz-Invocation-Type": "'Event'"
 }
```
Didn't test with real AWS API GW. Expected beheviour would that it return HTTP 202.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by eahefnawy** [comment]
Thanks for jumping on this @snurmine ... From the checklist above, it seems that this PR is ready for review, however it seems that the documention is missing. Could you add that?

> Didn't test with real AWS API GW. Expected beheviour would that it return HTTP 202.

Could you test it with real AWS API GW? :) ... Unit tests are great, but we can't merge anything that is untested manually.

File: lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js [hunk 1]
```diff
@@ -74,7 +74,13 @@ module.exports = {
             corsPreflight[http.path] = cors;
           }
 
-          http.integration = this.getIntegration(http, functionName);
+          // convert Servless only type AWS_ASYNC to
```
Comment (jplock) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/validate.js]: spelling: `Serverless`

**Comment by snurmine** [comment]
Agree this should be tested with real AWS API GW although I took hints from the issue.

**Comment by snurmine** [comment]
Tested with 

````
service: myService 

provider:
  name: aws
  runtime: nodejs8.10

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/create
          method: get
          integration: aws_async
````

`````
'use strict';

module.exports.hello = async (event) => {

 await ti(1000)

  return {
    statusCode: 400,
    body: JSON.stringify({
      message: 'Go Serverless v1.0! Your function executed successfully!',
      input: event,
    }),
  };

};

function ti(millis) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve('');
    }, millis)
  })
}
`````

It seems that it return before sleep is done, because it respons less than 1s.
if commented that  ` integration: aws_async` it takes more than 1s. 
Response code by default seems to be 200. It might require additional configuration in the template to get 202.

**Comment by eahefnawy** [comment]
Thanks for the updates @snurmine and for adding the docs. I'm gonna test this and get back to you! 🙌

**Comment by eahefnawy** [comment]
@snurmine works great. One more feedback though, don't you think it'd be better UX if we config it like this:

```yml
functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/create
          method: get
          async: true # default is false
```

rather than:

```yml
functions:
  hello:
    handler: handler.hello
    events:
      - http:
          path: users/create
          method: get
          integration: aws_async
```

This would give a nicer experience imo like we're doing with cors. It'd also match how you're validating it. What you think?

**Comment by snurmine** [comment]
Could be, since integration refers to aws apigw integration types(mock, lambda etc.). Only thing is that, maybe the feature is not so helpful. I wonder if something like adding custom headers in general, would be more beneficial. Then adding giving custom header and setting integration type to "lambda"(or what allows customs mapping) would give same effect.

**Comment by snurmine** [comment]
Thank you. I removed  type `aws_async`.  The `async: true` was internal helper that was not promoted in the template level. Just using `async: true` is cleaner.

**Comment by horike37** [comment]
Great! This is exactly what I just wanted 👍 
Thanks for implementing @snurmine

**Comment by bsdkurt** [comment]
Glad to see someone took this on and finished it!

I just wanted to point out that @HyperBrain mentioned this in comment https://github.com/serverless/serverless/issues/4862#issuecomment-376914187:

> As the asynchronous invocation is only valid with AWS type integrations, it could be encoded into a new integration type: `AWS_ASYNC`. Having a separate `async` property could make people believe that it is independent from the integration and should do something in any case.

Is `async: true` missleading for other integration types?

**Comment by snurmine** [comment]
Cloudformation might fail for integration types that don't support adding `RequestParameters`, but I didn't test. Could add some validation to definitely invalid combination, but it might be easier to let AWS validate the template. Now the integration type refers only to Cloudformation field `Integration.Type` and thus it might be easier to understand and existence of `async` just adds that extra header.

**Comment by louspringer** [comment]
Increasing the timeout over 30 seconds on an async lambda throws a warning like 
`Serverless: WARNING: Function post-ec2 has [a] timeout of 120 seconds, however, it's attached to API Gateway so it's automatically limited to 30 seconds.` 

Shouldn't the warning be suppressed for async lambdas?

**Comment by steelerc69** [comment]
How would the DLQ for the failures be configured for the async events?

**Pull request merged**